### PR TITLE
First pass at init function for SVCD

### PR DESCRIPTION
**Pull request by immesys**
This sets up the framework for the SVCD develoment (gradual replacement of symbols in the global SVCD table)

### CODE REVIEW

**Comment by gtfierro** [comment]
Looks good! Could use a couple more comments though, so the code is easier to follow.

Near the end of the `svcd_init` implementation, I'm guessing that [these lines](https://github.com/UCB-IoET/ioet_contrib/blob/b7f38764b4e667d8275d264e27d4de8ed6fbcfc9/natlib/svcd.c#L160-L183) correspond to the following function from [svcd.lua](https://github.com/UCB-IoET/ioet_contrib/blob/master/lib/svcd/svcd.lua)?

``` lua
storm.os.invokePeriodically(3*storm.os.SECOND, function()
    storm.net.sendto(SVCD.asock, storm.mp.pack(SVCD.manifest), "ff02::1", 2525)
end)
storm.bl.enable(id, SVCD.cchanged, onready)
```

Due to the obvious verbosity of implementing nested function calls (from modules, no less), it would not be a bad idea to add a few comments that inform the reader of what's on the stack and why these items are pushed in the particular order. No need to detail everything, but things like commenting which function is being executed with `lua_call` are definitely helpful

**Comment by gtfierro** [comment]
:ship: it

**Comment by samkumar** [comment]
Your code is readable, but it is quite repetitive near the beginning when you initialize all of the UDP sockets. You should consider creating a general function such as:

<pre><code>void make_socket(lua_State* L, int table_index, char* key_to_store, int port, char* func_name) {
    lua_pushstring(L, key_to_store);
    lua_pushlightfunction(L, libstorm_net_udpsocket);
    lua_pushnumber(L, port);
    lua_pushstring(L, func_name);
    lua_gettable(L, table_index);
    lua_call(L, 2, 1);
    lua_settable(L, 3);
}</code></pre>

And calling it for each UDP socket you create, instead of repeating the same structure of code over and over again. I haven't tested the above code; I included it to communicate the idea.
Applying the modification above will also make the code more concise.

**Comment by gtfierro** [comment]
Yes, this could be abstracted, but using the stack in this way acts in direct violation of how we told people to use the stack from C functions. Calling functions that operate on the stack from within another function means that you have to take special care to balance the stack before/after each call, which does not work well as a pedagogical device. Explicit code is more straightforward and easier to understand

**Comment by samkumar** [comment]
Gabe, I don't think I understand what you're saying. The function I wrote is stack balanced is it not? I thought it's OK to do this since I'm balancing the stack manually and I'm not relying on Lua to do it.

**Comment by gtfierro** [comment]
It is stack balanced, but the purpose of the code here is not to write the best possible DRY code, but rather to illustrate to people how they might go about writing code. Making something more complicated than it needs to be does not help people understand. When starting out with writing C extensions in Lua, it is best to avoid these types of functions because they can tend towards bugs.

**Comment by gtfierro** [comment]
To put it more succinctly, one can write a 1 or 2 line fizzbuzz program in Python, but if one is teaching someone else how to code, it's best to write it as simply as possible.

**Comment by calnaren** [comment]
I was slightly confused with the svcd_init_adv_received() function on line 34 of natlib/svcd.c. Is it supposed to be just to replace the custom function that is defined on line 75 of svcd.lua? How does lua_gettop() work when the arguments passed into the function is (...)?

**Comment by calnaren** [comment]
Also, I feel this is a good link to understand what each of the lua functions mean: http://www.lua.org/manual/5.1/manual.html

**Comment by samkumar** [comment]
Gabe, now I understand your point. Thanks for clarifying!

**Comment by immesys** [comment]
@calnaren lua_gettop() as the first thing in a function will tell you how many arguments you have (because those are pushed onto the stack for you) so for a varargs function, it is telling me how many arguments were passed. init_adv_received just loads the correct advert received function out the SVCD table and calls it with the same arguments that it received. We do this instead of passing the function directly to the socket, because it allows us to change the function that gets executed on advertisements being received later (same as in the Lua code)

File: natlib/svcd.c [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        lua_gettable(L, 3);
+        //Pack SVCD.manifest
+        lua_pushlightfunction(L, libmsgpack_mp_pack);
+        lua_pushstring(L, "manifest");
+        lua_gettable(L, 3);
+        lua_call(L, 1, 1);
+        //Address
+        lua_pushstring(L, "ff02::1");
+        lua_pushnumber(L, 2525);
+        cord_dump_stack(L);
```
Comment (leonardt) [inline natlib/svcd.c:174]: What is cord_dump_stack responsible for doing?

Comment (leonardt) [inline natlib/svcd.c:174]: Nevermind we found the implementation,
It is used for debugging (prints the entire stack to stdout)

**Pull request merged**